File: src/Backtracking/Exercise77.java
package Backtracking;

import java.util.ArrayList;
import java.util.List;

public class Exercise77 {
    class Solution {
        public List<List<Integer>> combine(int n, int k) {
            List<List<Integer>> ans = new ArrayList<>();
            combine(n, k, ans, new ArrayList<>(), 1);

            return ans;

        }

        private void combine(int n, int k, List<List<Integer>> ans, ArrayList<Integer> comb, int index) {

            if (comb.size() == k) {
                ArrayList<Integer> temp = new ArrayList<>(comb);
                ans.add(temp);
                return;
            }

            for (int i = index; i <= n; i++) {
                comb.add(i);
                combine(n, k, ans, comb, i + 1);
                comb.remove(comb.size() - 1);
            }
        }
    }
    
    public static void main(String[] args) {
        Exercise77 exercise = new Exercise77();
        Solution solution = exercise.new Solution();
        System.out.println(solution.combine(4, 2));

    }

}

File: src/Backtracking/Exercise77.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<vector<int>> combine(int n, int k)
    {
        vector<vector<int>> ans;
    }

private:
    vector<vector<int>> ans;
    void helper(int n, int start, int k, vector<int> &temp)
    {

        if (k == 1)
        {
            ans.push_back(temp);
            return;
        }

        temp.push_back(start);

        for (int i = start + 1; i <= n; i++)
        {
            helper(n, start + 1, k - 1, temp);
        }
    }
};